src/output/text.cpp

```diff
@@ -0,0 +1,604 @@
+  }
+};
+
+template <>
+struct TextEmitter<Value::OrderedMap> {
+  static void emit(std::ostream &out, const Value::OrderedMap &m)
+  {
+    for (const auto &[key, value] : m.values) {
+      out << "[";
+      TextEmitter<Primitive>::emit(out, key);
+      out << "]: ";
+      TextEmitter<Value>::emit(out, value);
+      out << std::endl;
+    }
+  }
+};
+
+template <>
+struct TextEmitter<Value::Stats> {
+  static void emit(std::ostream &out, const Value::Stats &s)
+  {
+    TextEmitter<std::decay_t<decltype(s.value)>>::emit(out, s.value);
+  }
+};
+
+template <typename T>
+static void try_add(Primitive &p)
+{
+  if (std::holds_alternative<T>(p.variant)) {
+    auto v = std::get<T>(p.variant);
+    if (v < std::numeric_limits<T>::max()) {
+      v += 1;
+    }
+  }
+}
+
+template <typename T>
+static void try_sub(Primitive &p)
+{
+  if (std::holds_alternative<T>(p.variant)) {
+    auto v = std::get<T>(p.variant);
+    if (v > std::numeric_limits<T>::min()) {
+      v -= 1;
+    }
+  }
+}
+
+static int64_t distance(const Primitive &p, int64_t other)
+{
+  if (std::holds_alternative<uint64_t>(p.variant)) {
+    const auto &v = std::get<uint64_t>(p.variant);
+    if (other < 0) {
+      return static_cast<uint64_t>(-other) + v;
+    }
+    if (v > static_cast<uint64_t>(other)) {
+      return static_cast<int64_t>(v - static_cast<uint64_t>(other));
+    } else {
+      return -static_cast<int64_t>(static_cast<uint64_t>(other) - v);
+    }
+  }
+  if (std::holds_alternative<int64_t>(p.variant)) {
+    const auto &v = std::get<int64_t>(p.variant);
+    return v - other;
+  }
+  return 0;
+}
+
+static int64_t distance(const Primitive &p, const Primitive &other)
+{
+  if (std::holds_alternative<int64_t>(other.variant)) {
+    return distance(p, std::get<int64_t>(other.variant));
+  } else if (std::holds_alternative<uint64_t>(other.variant)) {
+    return distance(p, static_cast<int64_t>(std::get<uint64_t>(other.variant)));
+  } else {
+    return 0; // Only handle integer primitives for now.
+  }
+}
+
+template <>
+struct TextEmitter<Value::TimeSeries> {
+  static void emit(std::ostream &out, const Value::TimeSeries &ts)
+  {
+    constexpr int graph_width = 53;
+
+    if (ts.values.empty()) {
+      out << "<no data>";
+      return;
+    }
+
+    std::vector<std::string> times;
+    output::Primitive min_value = ts.values.front().second;
+    output::Primitive max_value = ts.values.front().second;
+    for (const auto &[epoch, v] : ts.values) {
+      if (std::holds_alternative<std::monostate>(v.variant)) {
+        continue; // Skip missing values.
+      }
+      if (v < min_value) {
+        min_value = v;
+      }
+      if (v > max_value) {
+        max_value = v;
+      }
+    }
+    if (min_value == max_value) {
+      // Generally prefer if we can add some buffer on both sides of the
+      // points in the graph, but don't overflow if our min or max is at the
+      // upper bound of the range.
+      try_add<uint64_t>(max_value);
```

Comment:
nit: Personally I would prefer if the type check was open-coded at this level instead of being buried in `try_add`, as I think it makes the intent slightly clearer.

```cpp
  if (std::holds_alternative<uint64_t>(min_value)) {
    try_sub<uint64_t>(min_value);
  } else {
    try_sub<int64_t>(min_value);
  }

  if (std::holds_alternative<uint64_t>(max_value)) {
    try_add<uint64_t>(max_value);
  } else {
    try_add<int64_t>(max_value);
  }
```

or

```cpp
  if (std::holds_alternative<uint64_t>(min_value)) {
    try_add<uint64_t>(max_value);
    try_sub<uint64_t>(min_value);
  } else {
    try_add<int64_t>(max_value);
    try_sub<int64_t>(min_value);
  }
```

Reply:
Sure, I will do this on the next rebase & iteration.

Reply:
Looks nicer. Thanks.